cluster/connection_strategy_config.go

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cluster
+
+import (
+	"time"
+
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+type ConnectionStrategyConfig struct {
+	Retry            ConnectionRetryConfig
+	Timeout          types.Duration `json:",omitempty"`
+	DisableReconnect bool           `json:",omitempty"`
+}
+
+func (c ConnectionStrategyConfig) Clone() ConnectionStrategyConfig {
+	return c
+}
+
+func (c *ConnectionStrategyConfig) Validate() error {
+	if c.Timeout <= 0 {
+		// maximum duration
+		c.Timeout = types.Duration(1<<63 - 1)
+	}
+	return c.Retry.Validate()
+}
+
+type ConnectionRetryConfig struct {
+	InitialBackoff types.Duration `json:",omitempty"`
+	MaxBackoff     types.Duration `json:",omitempty"`
+	Multiplier     float64        `json:",omitempty"`
+	Jitter         float64        `json:",omitempty"`
+}
+
+func (c ConnectionRetryConfig) Clone() ConnectionRetryConfig {
+	return c
+}
+
+func (c *ConnectionRetryConfig) Validate() error {
+	if c.InitialBackoff <= 0 {
+		c.InitialBackoff = types.Duration(1 * time.Second)
+	}
+	if c.MaxBackoff <= 0 {
+		c.MaxBackoff = types.Duration(30 * time.Second)
+	}
+	if c.Multiplier <= 0 {
```

Comment:
Just a thought that comes to my mind after yesterday's review :smile: 

Should we change them such that if they are equal to zero value, set them to default value, if they are negative (or somehow out-of-valid-range), throw?

For example, for multiplier, values between 0 and 1.0 are not valid at all, as it would shorten the duration between backoffs, instead of increasing it. Should we throw on such cases?

This is not just related to this, but to whole configuration

Reply:
OK, here it is: https://github.com/hazelcast/hazelcast-go-client/pull/590/commits/d19d9f74ec76de7388c57eb1cbe0a5bc1e68a057